liblwgeom/lwgeom_geos.c

```diff
@@ -1223,13 +1227,97 @@ lwgeom_offsetcurve(const LWLINE* lwline, double size, int quadsegs, int joinStyl
 
 	g3 = GEOSOffsetCurve(g1, size, quadsegs, joinStyle, mitreLimit);
 
-	if (!g3) return geos_clean_and_fail(g1, NULL, NULL, __func__);
+	if (!g3)
+	{
+		geos_clean(g1, NULL, NULL);
+		return NULL;
+	}
 
 	if (!output_geos_as_lwgeom(&g3, &result, srid, is3d, __func__))
 		return geos_clean_and_fail(g1, NULL, g3, __func__);
 
 	geos_clean(g1, NULL, g3);
+	return result;
+}
+
+static LWGEOM*
+lwmline_offsetcurve(const LWGEOM* geom, double size, int quadsegs, int joinStyle, double mitreLimit)
+{
+	int32_t srid = get_result_srid(geom, NULL, __func__);
+	uint8_t is3d = FLAGS_GET_Z(geom->flags);
+	if (srid == SRID_INVALID) return NULL;
+
+	if (geom->type == LINETYPE)
+		return lwline_offsetcurve(lwgeom_as_lwline(geom), size, quadsegs, joinStyle, mitreLimit);
+	else if (geom->type == MULTILINETYPE || geom->type == COLLECTIONTYPE)
+	{
+		LWCOLLECTION* result;
+		LWGEOM* tmp;
+		const LWCOLLECTION* col = lwgeom_as_lwcollection(geom);
+		uint32_t i;
+		result = lwcollection_construct_empty(MULTILINETYPE, srid, is3d, LW_FALSE);
+		for (i = 0; i < col->ngeoms; i++)
+		{
+			tmp = lwmline_offsetcurve(col->geoms[i], size, quadsegs, joinStyle, mitreLimit);
+
+			if (!tmp)
+			{
+				lwcollection_free(result);
+				return NULL;
+			}
+
+			if (!lwgeom_is_empty(tmp))
+			{
+				if (lwgeom_is_collection(tmp))
+					result = lwcollection_concat_in_place(result, lwgeom_as_lwcollection(tmp));
+				else
+					result = lwcollection_add_lwgeom(result, tmp);
+
+				if (!result) {
+					lwgeom_free(tmp);
+					return NULL;
+				}
+			}
+		}
+
+		if (result->ngeoms == 1)
+		{
+			tmp = result->geoms[0];
+			lwcollection_release(result);
+			return tmp;
+		}
+		else
+			return lwcollection_as_lwgeom(result);
+	}
+	else
+		lwerror("%s: unsupported geometry type: %s", __func__, lwtype_name(geom->type));
+	return NULL;
+}
 
+LWGEOM*
+lwgeom_offsetcurve(const LWGEOM* geom, double size, int quadsegs, int joinStyle, double mitreLimit)
+{
+	int32_t srid = get_result_srid(geom, NULL, __func__);
+	LWGEOM* result;
+	if (srid == SRID_INVALID) return NULL;
+
+	if (lwgeom_dimension(geom) != 1)
+	{
+		lwerror("%s: input is not linear", __func__, lwtype_name(geom->type));
+		return NULL;
+	}
+
+	result = lwmline_offsetcurve(geom, size, quadsegs, joinStyle, mitreLimit);
+	if (!result)
+	{
```

Comment:
Is this going to print the `lwerror("%s: unsupported geometry type: %s", __func__, lwtype_name(geom->type));` from above?

Reply:
in context of postgres, lwerror is pg_error https://github.com/postgis/postgis/blob/c7e346cb4b4f04737c50f469b170ed288ab7c3b7/libpgcommon/lwgeom_pg.c#L121 which in turn calls `ereport(ERROR` which in turn cancels query at the spot (see https://www.postgresql.org/docs/10/static/error-message-reporting.html) - so if it prints that, it won't get to this stage.